calm_adapter/calm_api_client/src/main/scala/uk/ac/wellcome/platform/calm_api_client/CalmApiClient.scala

```diff
@@ -63,5 +99,25 @@ class CalmApiClient(
       case None         => request
     }
   }
+}
 
+class AkkaHttpCalmApiClient(
+  url: String,
+  username: String,
+  password: String,
+  minBackoff: FiniteDuration = 100 milliseconds,
+  maxBackoff: FiniteDuration = 30 seconds,
+  randomFactor: Double = 0.2,
+  maxRestarts: Int = 10
+)(implicit mat: Materializer)
+    extends HttpCalmApiClient(
+      url,
+      username,
+      password,
+      minBackoff,
+      maxBackoff,
+      randomFactor,
+      maxRestarts)
+    with AkkaHttpClient {
+  implicit val actorSystem: ActorSystem = mat.system
```

Comment:
Just checking my understanding as to why this is necessary - it's so we can test the `CalmApiClient` without having to worry about the AKKA stuff?

Reply:
Not so much akka stuff, it's so we can test it without making real HTTP requests

Reply:
I'll read through this again, but then why is it called `HttpCalmApiClient`?

Would it make sense to have the same pattern of `MemoryCalmApiClient` for testing and `HttpCalmApiClient` for when we interact with HTTP?

Reply:
Hmm, I don't think so - there are 2 use cases in test. One is where we don't care how the `CalmApiClient` works, and for that we use `TestCalmApiClient`, which is straightforward.

However, we have an implementor of `CalmApiClient` which is one that deals with `HttpResponse`s and `HttpRequest`s - that is therefore called `HttpCalmApiClient`.

It's non-trivial to do all the right things with this, so we want to test the client itself, but we don't really want to be doing actual HTTP requests because that becomes a headache.
So, we require that `HttpCalmApiClient`s must extend an `HttpClient`, which we can then use to provide responses without making requests in `TestHttpCalmApiClient` or we can use "for real" with `AkkaHttpCalmApiClient`.
This is actually very similar to how it was before (including naming), but using self-types like this makes it easier to mock in the right places than using constructor parameters like before.

Reply:
This makes more sense to me - thanks for running through it.

HTTP isn't using HTTP but is dealing with the `HTTPRequest/Response` types.